modules/core/src/main/java/org/apache/ignite/internal/managers/systemview/SystemViewArrayContainerAdapter.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.managers.systemview;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import org.apache.ignite.spi.systemview.view.SystemView;
+import org.apache.ignite.spi.systemview.view.SystemViewRowAttributeWalker;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * System view backed by {@code data} container.
+ * Each instance of {@code containers} collections should provide a array of data.
+ *
+ * @see SystemView
+ */
+public class SystemViewArrayContainerAdapter<C, R, D> extends AbstractSystemView<R> {
```

Comment:
Personally, I don't like this name, because "array container" means something storing the array. Here you don't store the array, but iterates through the nested array. But I'm ok with this name if there is no better name for the class.
The same for `SystemViewCollectionContainerAdapter` class and `register...` methods.

Reply:
Also, perhaps these two classes can be replaced by one which iterates through iterables. Size method doesn't really mandatory if you return `canGetRowCount` as `false` for these views. In this case size will be calculated by H2.

Reply:
Fixed